src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3472,68 +3493,125 @@ protected override void VisitStatement(BoundStatement statement)
 
         public override BoundNode? VisitCollectionExpression(BoundCollectionExpression node)
         {
-            // https://github.com/dotnet/roslyn/issues/68786: Use inferInitialObjectState() to set the initial
-            // state of the instance: see the call to InheritNullableStateOfTrackableStruct() in particular.
-            int containerSlot = GetOrCreatePlaceholderSlot(node);
-            NullableFlowState resultState = NullableFlowState.NotNull;
+            // When the collection is target-typed, we can initially only visit the elements and update the state.
+            // When the target-typing conversion is processed, the completion continuation will be given a target-type and
+            // we'll be able to process the element conversions and compute the final visit result.
 
-            var collectionKind = ConversionsBase.GetCollectionExpressionTypeKind(this.compilation, node.Type, out var targetElementType);
-            if (collectionKind is CollectionExpressionTypeKind.CollectionBuilder)
-            {
-                var createMethod = node.CollectionBuilderMethod;
-                if (createMethod is not null)
-                {
-                    var readOnlySpan = (NamedTypeSymbol)createMethod.Parameters[0].Type;
-                    targetElementType = readOnlySpan.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0];
-                    var annotations = createMethod.GetFlowAnalysisAnnotations();
-                    resultState = ApplyUnconditionalAnnotations(createMethod.ReturnTypeWithAnnotations, annotations).ToTypeWithState().State;
-                }
-            }
+            var (collectionKind, elementType) = getCollectionDetails(node, node.Type);
 
+            var resultBuilder = ArrayBuilder<VisitResult>.GetInstance(node.Elements.Length);
+            var elementConversionCompletions = ArrayBuilder<Func<TypeWithAnnotations, TypeWithState>>.GetInstance();
             foreach (var element in node.Elements)
             {
                 switch (element)
                 {
                     case BoundCollectionElementInitializer initializer:
+                        // We don't visit the Add methods
+                        // But we should check conversion to the iteration type (for that we need to have identified the preferred IEnumerable<T> interface)
```

Comment:
>(for that we need to have identified the preferred IEnumerable<T> interface) [](http://example.com/codeflow?start=80&length=76)

Consider removing this part of the comment since it will become stale if [https://github.com/dotnet/csharplang/pull/7783](https://github.com/dotnet/csharplang/pull/7783) is merged.